rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp

```diff
@@ -222,11 +227,9 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
       }
     }
 
-    for (auto service : service_handles_) {
-      if (rcl_wait_set_add_service(wait_set, service.get(), NULL) != RCL_RET_OK) {
-        RCUTILS_LOG_ERROR_NAMED(
-          "rclcpp",
-          "Couldn't add service to wait set: %s", rcl_get_error_string().str);
+    for (auto service_element : service_handles_) {
+      auto service = service_map_[service_element].lock();
+      if (service == nullptr || service->add_to_wait_set(wait_set) == false) {
         return false;
```

Comment:
definitely unrelated to your PR, but I find it strange that if we fail to add a service/client/subscription/timer then we return immediately, but we don't care if we failt to add a waitable or a guard condition.
Actions are waitables and they effectively include services and subscriptions inside.

Reply:
I think we can think about it from separated PR if needed. need to check whether there is somehow side effects due to throw the exception.

Reply:
Yes, definitely.
No need to make any changes as part of this PR.